src/storage/txn/scheduler.rs

```diff
@@ -154,109 +153,152 @@ pub struct Scheduler<E: Engine> {
     high_priority_pool: SchedPool<E>,
 
     // used to control write flow
-    running_write_bytes: usize,
+    running_write_bytes: AtomicUsize,
+}
+
+#[inline]
+fn id_index(cid: u64) -> usize {
+    cid as usize % TASKS_SLOTS_NUM
 }
 
+impl<E: Engine> SchedulerInner<E> {
+    /// Generates the next command ID.
+    #[inline]
+    fn gen_id(&self) -> u64 {
+        let id = self.id_alloc.fetch_add(1, Ordering::Relaxed);
+        id + 1
+    }
+
+    fn dequeue_task(&self, cid: u64) -> Task {
+        let mut tasks = self.task_contexts[id_index(cid)].lock().unwrap();
+        let task = tasks.get_mut(&cid).unwrap().task.take().unwrap();
+        assert_eq!(task.cid, cid);
+        task
+    }
+
+    fn enqueue_task(&self, task: Task, callback: StorageCb) {
+        let cid = task.cid;
+        let tctx = TaskContext::new(task, &self.latches, callback);
+
+        let running_write_bytes = self
+            .running_write_bytes
+            .fetch_add(tctx.write_bytes, Ordering::AcqRel) as i64;
+        SCHED_WRITING_BYTES_GAUGE.set(running_write_bytes + tctx.write_bytes as i64);
+        SCHED_CONTEX_GAUGE.inc();
+
+        let mut tasks = self.task_contexts[id_index(cid)].lock().unwrap();
+        if tasks.insert(cid, tctx).is_some() {
+            panic!("TaskContext cid={} shouldn't exist", cid);
+        }
+    }
+
+    fn dequeue_task_context(&self, cid: u64) -> TaskContext {
+        let tctx = self.task_contexts[id_index(cid)]
+            .lock()
+            .unwrap()
+            .remove(&cid)
+            .unwrap();
+
+        let running_write_bytes = self
+            .running_write_bytes
+            .fetch_sub(tctx.write_bytes, Ordering::AcqRel) as i64;
+        SCHED_WRITING_BYTES_GAUGE.set(running_write_bytes - tctx.write_bytes as i64);
+        SCHED_CONTEX_GAUGE.dec();
+
+        tctx
+    }
+
+    fn too_busy(&self) -> bool {
+        fail_point!("txn_scheduler_busy", |_| true);
+        self.running_write_bytes.load(Ordering::Acquire) >= self.sched_pending_write_threshold
+    }
+
+    /// Tries to acquire all the required latches for a command.
+    ///
+    /// Returns `Some(TaskContext)` if successful; returns `None` otherwise.
+    fn acquire_lock(&self, cid: u64) -> bool {
+        let mut task_contexts = self.task_contexts[id_index(cid)].lock().unwrap();
+        let tctx = task_contexts.get_mut(&cid).unwrap();
+        let acquired = self.latches.acquire(&mut tctx.lock, cid);
+        tctx.on_schedule();
```

Comment:
if not get the lock, still `on_schedule`? Also we can do it out of `acquire_lock` just like before, it is a little confusing that do schedule in `acquire_lock`
```
fn try_to_wake_up(&mut self, cid: u64) {
        let wake = if let Some(tctx) = self.acquire_lock(cid) {
            tctx.on_schedule();
            true
        } else {
            false
        };
        if wake {
            self.get_snapshot(cid);
        }
    }
```

Reply:
Now can't return a `&TaskContext` for a lifetime issue. However it's a nice catch about `on_schedule`, thanks.